File: lab3/oolab/src/main/java/agh/ics/oop/model/MapDirection.java
package agh.ics.oop.model;

import java.util.Map;

public enum MapDirection {
    NORTH (0),
    EAST (1),
    SOUTH (2),
    WEST (3);

    private final int value;

    MapDirection(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }

    @Override
    public String toString(){
        return switch(this){
            case NORTH -> "Północ";
            case EAST -> "Wschód";
            case SOUTH -> "Południe";
            case WEST -> "Zachód";
        };
    }

    public MapDirection next(){
        MapDirection[] directions = MapDirection.values();
        return directions[(this.getValue() + 1) % directions.length];
    }

    public MapDirection previous(){
        MapDirection[] directions = MapDirection.values();
        return directions[(this.getValue() + directions.length - 1) % directions.length];
    }

    public Vector2d toUnitVector(){
        int index = this.getValue();
        return new Vector2d((int) Math.pow(-1 * (index % 2), (int) index / 2 + 2), (int) Math.pow(-1 + index % 2, (int) index / 2 + 2));
    }
}

File: proj_lab2/oolab/src/main/kotlin/agh/ics/oop/model/MapDirection.kt
package agh.ics.oop.model

enum class MapDirection(private val value: Int) {
    NORTH(0),
    EAST(1),
    SOUTH(2),
    WEST(3);

    fun getValue(): Int = value

    override fun toString(): String = when (this) {
            NORTH -> "N"
            EAST -> "E"
            SOUTH -> "S"
            WEST -> "W"
    }

    fun next(): MapDirection {
        val directions = entries.toTypedArray()
        return directions[(value + 1) % directions.size]
    }

   fun previous(): MapDirection {
        val directions = entries.toTypedArray()
        return directions[(value + directions.size - 1) % directions.size]
    }

    //na podstawie punktu 2. z części 1. było:
    /*fun toUnitVector(): Vector2d = Vector2d(
            (-1 * (value % 2)).toDouble().pow((value / 2 + 2).toDouble()).toInt(),
            (-1 + value % 2).toDouble().pow((value / 2 + 2).toDouble()).toInt())*/
}
